### ISSUE TITLE
Call stack performance

### ISSUE THREAD

**Issue by brianhuffman** [GaloisInc/cryptol#1201]
TLDR: Basically, `combineCallStacks` is a performance bottleneck and we should try to optimize its performance.

I recently did a profiling run on an expensive cryptol computation (lots of polynomial div/mod operations). Here are the top cost centers:
```
COST CENTRE                        MODULE                    %time %alloc

==                                 Cryptol.Parser.Position    18.9    0.0
combineCallStacks.dropCommonPrefix Cryptol.Backend.Monad      17.0   24.2
lookupSeqMap                       Cryptol.Backend.SeqMap     11.4    5.5
evalExpr                           Cryptol.Eval                9.7   27.0
evalExpr->EVar                     Cryptol.Eval                4.8    1.8
evalExpr.\                         Cryptol.Eval                3.5   11.1
evalExpr->EApp                     Cryptol.Eval                3.4   11.8
evalComp.\                         Cryptol.Eval                2.4    2.1
```
The `(==)` operation from `Cryptol.Parser.Position` is actually called by `combineCallStacks.dropCommonPrefix`, so all together, `combineCallStacks` is taking up at least 35.9% of the total runtime, and 24.2% of the total allocation of the entire computation.

Here's the place where `combineCallStacks` is called the most in my use case:
```
COST CENTRE                                   MODULE                         entries  %time %alloc   %time %alloc

Prelude::(#)                                  Cryptol.Eval.Generic                 0    0.0    0.0    37.1   22.3
 genericPrimTable.\                           Cryptol.Eval.Generic                 0    0.0    0.0    37.1   22.3
  genericPrimTable.\.\                        Cryptol.Eval.Generic                 0    0.0    0.0    37.1   22.3
   genericPrimTable.\.\.\                     Cryptol.Eval.Generic                 0    0.0    0.0    37.1   22.3
    genericPrimTable.\.\.\.\                  Cryptol.Eval.Generic                 0    0.0    0.0    37.1   22.3
     genericPrimTable.\.\.\.\.\               Cryptol.Eval.Generic                 0    0.1    0.0    37.1   22.3
      lookupSeqMap                            Cryptol.Backend.SeqMap         9977247    8.0    2.8    36.7   22.3
       fromVFun.\.\                           Cryptol.Eval.Value             9949023    0.0    0.0    21.7   15.1
        combineCallStacks                     Cryptol.Backend.Monad          9949023    0.5    1.3    21.6   15.1
         combineCallStacks.dropCommonPrefix   Cryptol.Backend.Monad        113343321    8.8   13.8    21.1   13.8
          ==                                  Cryptol.Parser.Position      103394298   10.6    0.0    10.7    0.0
           ==                                 Cryptol.Parser.Position      206788596    0.1    0.0     0.1    0.0
          ==                                  Cryptol.ModuleSystem.Name    103394298    0.8    0.0     1.6    0.0
           compare                            Cryptol.ModuleSystem.Name    103394298    0.8    0.0     0.8    0.0
```
We can see that of the total time/allocation used for evaluating the `Prelude::(#)` primitive,  more than half of the runtime and 2/3 of the allocation is due to `combineCallStacks` alone.

As we can see in the code, the `dropCommonPrefix` function is calling `(==)` on type `(Name, Range)`. Comparing the `Name` alone is efficient because each `Name` contains a unique integer ID; but I guess just comparing `Name` is not enough? It would help to use a representation of `Range`s that can be compared more efficiently.

https://github.com/GaloisInc/cryptol/blob/bd8588adcaa6a6172c06359d8a394952d45cb7b7/src/Cryptol/Backend/Monad.hs#L66

https://github.com/GaloisInc/cryptol/blob/bd8588adcaa6a6172c06359d8a394952d45cb7b7/src/Cryptol/Backend/Monad.hs#L97-L101

The profiling also shows that `combineCallStacks` is called a very large number of times (almost 10 million, with over 100 million recursive calls to `dropCommonPrefix`). Surely many of these calls are with exactly the same arguments over and over again. I wonder if we can avoid this recomputation somehow?

**Comment by robdockins** [comment]
Obviously, it would be really nice to improve the implementation of call stacks, but for any performance sensitive Cryptol, I strongly recommend using the `--no-call-stacks` option for this reason.

**Comment by brianhuffman** [comment]
With #1202, I'm seeing something like a 40% speedup on some of my examples (running without profiling). It's still not nearly as fast as `--no-call-stacks`, so I should probably be using that anyway.

**Comment by weaversa** [comment]
Sorry for the slight derailment, but what does `cryptol-remote-api` do with call stacks? It doesn't have this flag.

**Comment by robdockins** [comment]
The cryptol server currently hard-codes this option so that it will not track call stacks.  I'm not sure what's the right way to configure this, maybe the server should also have a command line option.

**Comment by robdockins** [comment]
Maybe we can do something better here, but I'm not convinced it is worth spending a lot of time on.

What we probably should do, however, is make a note somewhere more prominent of the `--no-call-stacks` option; perhaps in the book.  Any ideas where would be the best place to put something like that?  Maybe we need a separate document or new chapter in the book on how to tackle performance issues...
